grants_tagger/tag_grants.py

```diff
@@ -8,63 +8,51 @@
-        tagged_grant = {
-            'Grant ID': grant['grant_id'],
-            'Reference': grant['reference'],
-            'Grant No.': grant['grant_no']
-        }
-        tagged_grant.update({
-            f"Tag #{i+1}": tag
-            for i, tag in enumerate(tags)
-            if i < 10 
-        })
-        yield tagged_grant
-
-
-def tag_grants(grants_path, tagged_grants_path, model_path, label_binarizer_path, approach, threshold=0.5):
-    with open(tagged_grants_path, 'w') as f_o:
-        fieldnames = ["Grant ID", "Reference", "Grant No."]
-        fieldnames += [f"Tag #{i}" for i in range(1,11)]
-        csv_writer = csv.DictWriter(f_o, fieldnames=fieldnames)
+def yield_batched_grants(input_file, batch_size=256):
+    """yield grants in batches from input file line by line"""
+    with open(input_file, "r") as tf:
+        csv_reader = csv.DictReader(tf, delimiter=",", quotechar='"')
+        lines = []
+        for line in csv_reader:
+            lines.append(line)
+            if len(lines) >= batch_size:
+                yield lines
+                lines = []
+        if lines:
+            yield lines
+
+
+def tag_grants(grants_path, tagged_grants_path, model_path, label_binarizer_path, approach, threshold=0.5,
+        grant_id_field = "grant_id", grant_text_fields=["title", "synopsis"], text_null_value="No Data Entered"):
+
+    with open(tagged_grants_path, "w") as tagged_grants_tf:
+        fieldnames = ["Grant id", "Tag", "Prob"]
+        csv_writer = csv.DictWriter(
+            tagged_grants_tf, fieldnames=fieldnames, delimiter=",",
+            quotechar='"', quoting=csv.QUOTE_MINIMAL
+        )
         csv_writer.writeheader()
 
-        for grants in yield_grants(grants_path, batch_size=512):
-            for tagged_grant in yield_tagged_grants(grants, model_path, label_binarizer_path, approach, threshold):
-                csv_writer.writerow(tagged_grant)
+        for grants in yield_batched_grants(grants_path, 128):
+            grants_text = [
+                " ".join([
+                    grant[field].replace(text_null_value, "")
+                    for field in grant_text_fields
+                ])
+                for grant in grants
+            ]
+
+            # Removes consecutive white spaces which are uninformative and may cause error #30
+            grants_text = [" ".join(text.split()) for text in grants_text if text.strip()] # Removes empty text
+            grants_tags = predict_tags(grants_text, model_path, label_binarizer_path, approach,
+                    probabilities=True, threshold=threshold)
+
+            for grant, tags in zip(grants, grants_tags):
+                for tag, prob in tags.items():
+                    csv_writer.writerow({
+                        'Grant id': grant[grant_id_field],
+                        'Tag': tag,
+                        'Prob': prob
+                    })
+
+            tagged_grants_tf.flush()
```

Comment:
do you need this? is it something to do with the way csv_writer handles I/O?

Reply:
tbh I copied and pasted from https://github.com/wellcometrust/data-pipelines/blob/739534868294dd0047934e859800a800f5c5c3bd/pipelines/mesh/task_tagger.py#L81

maybe @jdu can enlighten us if this is needed

Reply:
Basically, when you write to a file, in this case calling `writerow` it doesn't actually write to the file immediately, it writes to a buffer. That buffer gets "flushed" to the file periodically. What we're doing here by explicitly calling `flush` is making sure that every time we start another batch of 128 grants, we flush the results to disk. 

It reduces memory overhead so you're not carrying around an ever-growing buffer for a long period of time. 

Usually when you use something like

```
with open("somefile.csv", "wb") as f:
    f.writerow([])
```

python automatically flushes the buffer to disk once you leave the `with` block, but with a longer running process like grant tagger, mesh tagger, etc... where we want to flush that buffer manually periodically to keep our memory usage down.

Reply:
thanks for the explanation, makes sense

Reply:
Nice! Thanks, I didn't know that about `writerow`!